### PR TITLE
Address max sequence length issue (#1002)

### PR DESCRIPTION
**Pull request by pyeres**
This PR proposes a change to addresses the max_seq_len issue: https://github.com/nyu-mll/jiant/issues/1002. 

These changes were validated on [the example](https://github.com/nyu-mll/jiant/issues/1002#issue-559833991) provided by @zphang. @W4ngatang for visibility.

### CODE REVIEW

File: jiant/preprocess.py [hunk 1]
```diff
@@ -785,11 +870,15 @@ def __init__(self, args):
         else:
             boundary_token_fn = utils.apply_standard_boundary_tokens
 
-        self.boundary_token_fn = boundary_token_fn
+        # TODO: consider the max_tokens from a model-specific config instead of args.max_seq_len?
```
Comment (W4ngatang) [inline jiant/preprocess.py]: probably should be something like `min(args.max_seq_len, model.max_pos)` (assuming `args.max_seq_len` is set, and defaulting to `model.max_pos` if not)

Comment (pyeres) [reply jiant/preprocess.py]: 👍, done w/ 65912cc

**Comment by pyeres** [comment]
Thanks for your notes, @W4ngatang —

> This makes sense to me! Scattered thoughts:
> 
> * AFAICT, there's no way to pass `trunc_strategy` right now

Not directly, but the outer method `_apply_boundary_tokens` returns the inner function `_apply_boundary_tokens_with_trunc_strategy`, exposing it to be called with keyword args.

> * will this break if there's a single sentence task that's really long w/ `trunc_strategy = 'trunc_s2'` as the default? I think so b/c it will try to access `args[1]`?
> * I can imagine some pretty subtle bugs arising based on fields being truncated unintentionally based on how the example fields are ordering (e.g. if I've reversed my question and my paragraph, I may cut off a lot of may question). Do you think there's a clean way to do this with named fields, or just rely on lots of loud warnings in the docs?

1) To simplify things I'm updating all the existing `apply_boundary_tokens` functions to require that `s2` be passed as a keyword arg (this is the existing signature, but the existing signature also allowed passing `s2` as a positional arg which made the wrapper logic prone to bugs).
2) I'm updating all the callers of `apply_boundary_tokens` to explicitly provide `s2` as a keyword.
3) To avoid truncating anything implicitly, these changes will now raise an exception when `apply_boundary_tokens` is called IF the sequence is longer than the model/max_seq_len allows AND the truncation strategy was left unspecified. This will push task-related decisions about how to truncate back to the task.

> * We'll probably need to update the docs to point out to people implementing new tasks how to write this and what the strategy should be set to for different tasks, though I think can just truncate s2.

Agreed, happy to turn to this (and finalize the docstrings) if we're aligned on design.

File: jiant/preprocess.py [hunk 2]
```diff
@@ -743,53 +743,171 @@ class ModelPreprocessingInterface(object):
...
+            seq_w_boundry_tokens = apply_boundary_tokens(*args, **kwargs)
+            # if after calling the apply_boundary_tokens function the number of tokens is greater
+            # than the model's max, a truncation strategy can reduce the number of tokens:
+            num_excess_tokens = len(seq_w_boundry_tokens) - max_tokens
+            if num_excess_tokens > 0:
+                if not (trunc_strategy and trunc_side):
+                    raise ValueError(
+                        "Input(s) will exceed model capacity or max_seq_len. "
+                        + "Adjust settings as necessary, or to automatically truncate "
+                        + "inputs in apply_boundary_tokens, call apply_boundary_tokens"
```
Comment (W4ngatang) [inline jiant/preprocess.py]: + space

File: jiant/preprocess.py [hunk 3]
```diff
@@ -743,53 +743,171 @@ class ModelPreprocessingInterface(object):
...
+                    + str(max_tokens)
+                    + ". Truncation strategy "
+                    + trunc_strategy
+                    + " will be applied, removing elements from the "
+                    + trunc_side
+                    + " side of the input."
+                )
+                if trunc_strategy == "trunc_s2":
+                    s2 = kwargs["s2"]
+                    log.warning("Before truncation, s2 length = " + str(len(s2)))
```
Comment (W4ngatang) [inline jiant/preprocess.py]: incredibly inconsequential: I'd prefer to just print the before/after length on the same line. Some datasets might have lots of lines to be truncated.

Comment (pyeres) [reply jiant/preprocess.py]: Sounds like a good idea to me — done in 09cb503.

**Comment by pyeres** [comment]
At this point the PR implements a mechanism for applying truncation strategies which are specifiable when tasks call the `model_preprocessing_interface`. But actually specifying default strategies for the tasks isn't done yet. Planning to request info on reasonable defaults from the user group.

**Comment by pyeres** [comment]
@W4ngatang, @zphang, @sleepinyourhat — request for comment:

Should tasks' strategies for handling model inputs be specified in the task's code, or in the config?

If in the task's code, the implementation in this PR requires specifying a strategy in any task that need a strategy (and some might not need a strategy because their inputs are small).

If in the config, we need a way to pass that option to the task. One way to do that would be to use the route I suggested in #1003.

What do you think?

**Comment by sleepinyourhat** [comment]
I think that truncation strategy will rarely vary within a task—the only case where that might occur is for a very generic task that is @register-ed to point to very different datasets. I think it will basically never vary across runs in a single experiment.

So, I think it's safe/reasonable to leave it as a hardcoded property of task objects if that's expedient. It'd be nice to expose this as an optional task-specific config override, but I think that'll be useful rarely enough that it's safe to merge in this change without that exposure.

**Comment by W4ngatang** [comment]
I agree with Sam. Leaving the truncation strategy hardcoded in the task class will cover almost all of the cases. Also, we discussed this in person, but I'd advocate for an implementation of an `equal` strategy, where `s1` and `s2` are equally truncated.

I am now realizing there's some weirdness in that we are now truncating in two places: once when we load the raw text data and again when using models that concatenate all inputs. Is it worth restructuring so that we only truncate once (probably not, as this would almost never be significant)?

**Comment by pyeres** [comment]
> I am now realizing there's some weirdness in that we are now truncating in two places: once when we load the raw text data and again when using models that concatenate all inputs. Is it worth restructuring so that we only truncate once (probably not, as this would almost never be significant)?

From a performance/correctness point of view, it doesn't seem like a problem to sometimes truncate more than once. However, I agree that it would be good to reduce unnecessary complexity.

IINM, the first instance of truncation that you're referring to is in the `tokenize_and_truncate` function (`/utils/data_loaders.py`), and this fn has a naive approach to account for boundary tokens (`max_seq_len -= 2`), and at this point in the code which tokenizer to apply is known (so perhaps we can use a less naive method). 

But later on, when the tasks' `process_split` method is called, a good deal of conditional logic is used to decide if/how to concatenate inputs to form examples for the model. If this logic forms examples longer than model limits, there will be an error. So, it seems like it's possible to concat two inputs that were individually under the length limit at data loading time, but put together at the point of `process_split` could exceed the limit.

So, if my understanding of the above is correct, it seems like a good reason to have the truncation happen at/after the model-specific decisions on if/how to concatenate inputs (which is how the PR would arrange it). It's also possible that we can a) move this logic upstream to data loading time, or b) eliminate the truncation from `tokenize_and_truncate` and let it happen when the model interface is in play, but I think these would be more substantial refactors.

**Comment by pyeres** [comment]
> I agree with Sam. Leaving the truncation strategy hardcoded in the task class will cover almost all of the cases.

Commit 4cc77fc adds default truncation strategies for most common tasks.

A request was raised yesterday to add handling for any number of inputs (s_n) to `apply_boundary_tokens`. This will change the signature of the call to `apply_boundary_tokens`, but it shouldn't change the default truncation strategies. So, please take a look at the default strategies specified in 4cc77fc and let me know if they seem reasonable to you.

File: tests/test_preprocess.py [hunk 4]
```diff
@@ -75,3 +81,92 @@ def test_build_vocab(self):
...
+        seq = mpi.boundary_token_fn(
+            ["Apple", "buy", "call"],
+            s2=["Xray", "you", "zoo"],
+            trunc_strategy="trunc_s1",
+            trunc_side="left",
+        )
+        self.assertEqual(len(seq), MAX_SEQ_LEN)
+        self.assertEqual(seq, ["[CLS]", "call", "[SEP]", "Xray", "you", "zoo", "[SEP]"])
+
+    def test_boundary_token_fn_trunc_s1_left_side(self):
```
Comment (W4ngatang) [inline tests/test_preprocess.py]: duplicate test?

File: tests/test_preprocess.py [hunk 5]
```diff
@@ -75,3 +81,92 @@ def test_build_vocab(self):
...
+        seq = mpi.boundary_token_fn(
+            ["Apple", "buy", "call"],
+            s2=["Xray", "you", "zoo"],
+            trunc_strategy="trunc_both",
+            trunc_side="left",
+        )
+        self.assertEqual(len(seq), MAX_SEQ_LEN)
+        self.assertEqual(seq, ["[CLS]", "buy", "call", "[SEP]", "you", "zoo", "[SEP]"])
+
+    def test_boundary_token_fn_trunc_with_short_sequence_an_no_max_seq_len(self):
```
Comment (W4ngatang) [inline tests/test_preprocess.py]: nit: "an" -> "and"

File: tests/test_preprocess.py (hunk 4 above)
Comment (W4ngatang) [reply tests/test_preprocess.py]: I'm guessing this should left -> right?

File: jiant/preprocess.py [hunk 6]
```diff
@@ -743,53 +743,202 @@ class ModelPreprocessingInterface(object):
...
         elif args.input_module.startswith("xlm-"):
             from jiant.huggingface_transformers_interface.modules import XLMEmbedderModule
 
             boundary_token_fn = XLMEmbedderModule.apply_boundary_tokens
+            max_pos = XLMTokenizer.max_model_input_sizes.get(args.input_module, None)
         else:
             boundary_token_fn = utils.apply_standard_boundary_tokens
 
-        self.boundary_token_fn = boundary_token_fn
+        self.max_tokens = min(x for x in [max_pos, args.max_seq_len] if x is not None)
```
Comment (W4ngatang) [inline jiant/preprocess.py]: probably would never happen, but is it possible to get both as `None`? (though I think models always have `max_pos != None`)

**Comment by W4ngatang** [comment]
I'd prefer (b) and not (a), but I agree it'd be a larger refactor beyond this PR. In any case, seems low priority.

File: jiant/preprocess.py (hunk 6 above)
Comment (pyeres) [reply jiant/preprocess.py]: If `args.max_seq_len` isn't set, earlier calls to `tokenize_and_truncate` will raise a TypeError. So we shouldn't encounter a scenario where both are `None` here.

File: tests/test_preprocess.py (hunk 4 above)
Comment (pyeres) [reply tests/test_preprocess.py]: Thanks — yeah, should have been `right`. Fixing with next commit.

**Comment by pyeres** [comment]
> LGTM modulo some minor cleanup (see comments). I think the defaults for all tasks is reasonable/correct, though I'm not familiar with the format of Winogrande and Abductive NLI. I think `trunc_s1` and `trunc_right` are both reasonable.

Thanks @W4ngatang. For Abductive NLI I'll be conservative and check w/ author @zphang.

**Comment by pyeres** [comment]
> Thanks @W4ngatang. For Abductive NLI I'll be conservative and check w/ author @zphang.

Connected with @zphang — confirmed that to properly apply truncation strategy to `AlphaNLITask` (`(obs1 + hyp, s2=obs2)`), `obs1` should be truncated. This isn't handled by current logic, but has been raised as a feature request and is captured in #1009.

**Comment by pyeres** [comment]
After discussion with @sleepinyourhat, merging these changes to handle the most common cases. Leaving original issue (#1002) open to motivate followup work described in #1009.

**Pull request merged**